lib/rubocop/cop/style/redundant_regexp_argument.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # Identifies places where argument can be replaced from
+      # a deterministic regexp to a string.
+      #
+      # @example
+      #   # bad
+      #   'foo'.byteindex(/f/)
+      #   'foo'.byterindex(/f/)
+      #   'foo'.gsub(/f/, 'x')
+      #   'foo'.gsub!(/f/, 'x')
+      #   'foo'.index(/f/)
+      #   'foo'.partition(/f/)
+      #   'foo'.rindex(/f/)
+      #   'foo'.rpartition(/f/)
+      #   'foo'.scan(/f/)
+      #   'foo'.split(/f/)
+      #   'foo'.sub(/f/, 'x')
+      #   'foo'.sub!(/f/, 'x')
+      #
+      #   # good
+      #   'foo'.byteindex('f')
+      #   'foo'.byterindex('f')
+      #   'foo'.gsub('f', 'x')
+      #   'foo'.gsub!('f', 'x')
+      #   'foo'.index('f')
+      #   'foo'.partition('f')
+      #   'foo'.rindex('f')
+      #   'foo'.rpartition('f')
+      #   'foo'.scan('f')
+      #   'foo'.split('f')
+      #   'foo'.sub('f', 'x')
+      #   'foo'.sub!('f', 'x')
+      class RedundantRegexpArgument < Base
+        extend AutoCorrector
+
+        MSG = 'Use string `%<prefer>s` as argument instead of regexp `%<current>s`.'
+        RESTRICT_ON_SEND = %i[
+          byteindex byterindex gsub gsub! index partition rindex rpartition scan split sub sub!
+        ].freeze
+        DETERMINISTIC_REGEX = /\A(?:#{LITERAL_REGEX})+\Z/.freeze
```

Comment:
it is maybe not directly relevant to this PR as it already existed before, but the `LITERAL_REGEX` has both false positives and false negatives.

```ruby
# current regex
LITERAL_REGEX = %r{[\w\s\-,"'!#%&<>=;:`~/]|\\[^AbBdDgGhHkpPRwWXsSzZ0-9]}

# some false positives
LITERAL_REGEX.match?(/foo # comment/x.source) # => true
LITERAL_REGEX.match?(/a\Kb/.source) # => true (\K is a lookbehind)

# some false negatives
LITERAL_REGEX.match?(/😃/.source) # => false
LITERAL_REGEX.match?(/]/.source) # => false
LITERAL_REGEX.match?(/@/.source) # => false
LITERAL_REGEX.match?(/ü/.source) # => false
```

it might make sense to improve `LITERAL_REGEX` a bit, or maybe `regexp_parser` could be used, approximately like so (i haven't tested this):

```ruby
Regexp::Parser.parse(regexp).each_expression.all? do |(exp)|
  exp.one_of?(:literal, :escape) && !exp.quantified?
end
```

Reply:
`/foo # comment/` does not seem to be a false positive. OTOH, regular expression `/a\Kb/` may indeed be a false positive, or an incorrect autocorrection. `LITERAL_REGEXP` is a constant used in some existing RuboCop Performances cops, so I would like to modify it separately.

Reply:
> /foo # comment/ does not seem to be a false positive

It is not a false positive in the sense that it is impossible to replace it with a string, but IIUC the cop would replace it with `'foo # comment'`, which would not behave the same?

Reply:
For example, the following are compatible. What code specifically causes the problem?

```ruby
'foo # comment extra'.match?(/foo # comment/) # => true
'foo # comment extra'.match?('foo # comment') # => true
```

Reply:
`/foo # comment/x`

Reply:
also `/ foo /x`

Reply:
Ah, sure! It changed to allow when options are present.